### PR TITLE
Release microsoft.api management 2023 03 01 preview

### PR DESCRIPTION
**Pull request by solankisamir**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by zedy-wj** [comment]
@raych1 - Please help review Go breaking change.
@qiaozha - Please help review JS breaking change.

**Comment by TimLovellSmith** [comment]
Removing ARMSignedOff for now since it looks like there are further changes coming in after signoff.

**Comment by raosuhas** [comment]
@solankisamir please only remove the ARMCHangesRequested label once this PR is ready for ARM review to avoid multiple review iterations

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimdeployment.json [hunk 1]
```diff
@@ -0,0 +1,2113 @@
...
+              },
+              {
+                "value": "stv2.1",
+                "description": "Platform running the service on Single Tenant V2 platform on newer Hardware."
+              }
+            ]
+          }
+        },
+        "legacyPortalEnabled": {
+          "type": "boolean",
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimdeployment.json:1701]: >boolean [](http://example.com/codeflow?start=19&length=7)

consider using enum instead:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/enum-instead-of-boolean.md#why-the-rule-is-important](https://github.com/Azure/azure-openapi-validator/blob/main/docs/enum-instead-of-boolean.md#why-the-rule-is-important)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimdeployment.json [hunk 2]
```diff
@@ -0,0 +1,2113 @@
...
+          "description": "Enable legacy portal for this API Management service.",
+          "default": true
+        },
+        "developerPortalEnabled": {
+          "type": "boolean",
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimdeployment.json:1706]: >boolean [](http://example.com/codeflow?start=19&length=7)

consider using enum instead:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/enum-instead-of-boolean.md#why-the-rule-is-important](https://github.com/Azure/azure-openapi-validator/blob/main/docs/enum-instead-of-boolean.md#why-the-rule-is-important)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json [hunk 3]
```diff
@@ -0,0 +1,1473 @@
...
+        "operationId": "Gateway_InvalidateDebugCredentials",
+        "description": "Action is invalidating all debug credentials issued for gateway.",
+        "x-ms-examples": {
+          "ApiManagementGatewayInvalidateDebugCredentials": {
+            "$ref": "./examples/ApiManagementGatewayInvalidateDebugCredentials.json"
+          }
+        },
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json:1336]: >v3 [](http://example.com/codeflow?start=69&length=2)

please always use the latest, current is "v5". please also update other places

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json [hunk 4]
```diff
@@ -0,0 +1,1473 @@
...
+          "200": {
+            "description": "The response body contains debug credentials to use in gateway.",
+            "schema": {
+              "$ref": "./definitions.json#/definitions/GatewayDebugCredentialsContract"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "./apimanagement.json#/definitions/ErrorResponse"
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json:1411]: >"./apimanagement.json#/definitions/ErrorResponse" [](http://example.com/codeflow?start=22&length=49)

use ../../../../../common-types/resource-management/v5/types.json#/definitions/ErrorResponse if possible. please also update other places

**Comment by ms-zhenhua** [comment]
>       "additionalProperties": true,

additionalProperties is not allowed and boolean is only supported in 3.0:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/definitions.json:8857 in 63be757. [](commit_id = 63be75748a2709cb4f7cf787b468b8db76038b52, deletion_comment = False)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json [hunk 5]
```diff
@@ -0,0 +1,1473 @@
...
+            "$ref": "./apimanagement.json#/parameters/ServiceNameParameter"
+          },
+          {
+            "$ref": "./apimanagement.json#/parameters/GatewayIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "name": "parameters",
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json:1393]: >"name": "parameters", [](http://example.com/codeflow?start=12&length=21)

please add "description"

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json [hunk 6]
```diff
@@ -0,0 +1,1473 @@
...
```
Comment (ms-zhenhua) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json:1446]: >"name": "parameters", [](http://example.com/codeflow?start=12&length=21)

please add "description"

Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/apimgateways.json:1446]: fixed

**Comment by solankisamir** [comment]
@ms-zhenhua I have resolved the issues you raised.
For `additionalproperties` we will consider this for future api-versions. Can you review again?

**Comment by ms-henglu** [comment]
Hi @solankisamir ,

Would you please clarify why the following issue couldn't be fixed this time? 

> > ```
> >   "additionalProperties": true,
> > ```
> 
> 
> additionalProperties is not allowed and boolean is only supported in 3.0:
> https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md
> 
> 
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/definitions.json:8857 in [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52). [](commit_id = [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52), deletion_comment = False)

**Comment by solankisamir** [comment]
> Hi @solankisamir ,
> 
> Would you please clarify why the following issue couldn't be fixed this time?
> 
> > > ```
> > >   "additionalProperties": true,
> > > ```
> > 
> > 
> > additionalProperties is not allowed and boolean is only supported in 3.0:
> > https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md
> > Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/definitions.json:8857 in [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52). [](commit_id = [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52), deletion_comment = False)

@Mielek can you please clarify this?

**Comment by Mielek** [comment]
Hello @ms-henglu and @solankisamir 

We want to leave an open ended contract for the values of tracing. We made a conscious decision about that.

Sort explanation. Traces are tight to policy documents used in APIM. Policy documents contain policies and expressions. Policy can produce part of the traces. Expressions (customer defined scripts) can produce expressions. Every document can have different set of policies and expressions. If we would define the structure of trace then:
1. Every modification, improvement or addition of traces would require an update in ARM spec. We want to avoid that.
2. The trace structure can be much different for every policy document. it is because customer will use different set of policies, policies in different order and customer can add trace elements with unknow for us structure through expression.
3. It is a debug functionality. Customer, when it faces some issues in policy documents, will produce a trace. The trace will be checked manually to find any issues. We do not believe that we or customers will be making tooling around that functionality.

I hope it helps you to understand why we want to have an open ended contract here. Let me know if you have more questions around it.

> > Hi @solankisamir ,
> > Would you please clarify why the following issue couldn't be fixed this time?
> > > > ```
> > > >   "additionalProperties": true,
> > > > ```
> > > 
> > > 
> > > additionalProperties is not allowed and boolean is only supported in 3.0:
> > > https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md
> > > Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2023-03-01-preview/definitions.json:8857 in [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52). [](commit_id = [63be757](https://github.com/Azure/azure-rest-api-specs/commit/63be75748a2709cb4f7cf787b468b8db76038b52), deletion_comment = False)
> 
> @Mielek can you please clarify this?

Hello @ms-henglu and @solankisamir 

We want to leave an open ended contract for the values of tracing. We made a conscious decision about that.

Shortexplanation. Traces are tight to policy documents used in APIM. Policy documents contain policies and expressions. Policies can produce part of the traces. Expressions (customer defined scripts) can produce traces. Every document can have different set of policies and expressions. If we would define the structure of trace then:
1. Every modification, improvement or addition of traces would require an update in ARM spec. We treat trace as an event log. We do not want to block the addition of new log entries because of api-version. We as well do not see a point in filtering traces to remove the elements  which where not present in the api-version.
2. The trace structure can be much different for every policy document. It happens because customer will use different set of policies, policies in different order and customer can add trace elements with unknow for us structure through expression.
3. It is a debug functionality. Customer, when it faces some issues in policy documents, will produce a trace. The trace will be checked manually to find any issues. We do not believe that we or customers will be making tooling around that functionality.

I hope it helps you to understand why we want to have an open ended contract here. Let me know if you have more questions around it.

**Comment by ms-henglu** [comment]
Hi @Mielek ,

Thanks for the clarification, I've signed off this PR.

**Review by zedy-wj — APPROVED** [review]

**Comment by JackTn** [comment]
> @raych1 - Please help review Go breaking change. @qiaozha - Please help review JS breaking change.

@raych1  @qiaozha take a look on this

**Pull request merged**